src/content/en/tools/workbox/guides/advanced-recipes.md

```diff
@@ -0,0 +1,173 @@
+  }
+
+  const listenInstalledStateChange = () => {
+    registration.installing.addEventListener('statechange', () => {
+      if (event.target.state === 'installed') {
+        // A new service worker is available, inform the user
+        callback();
+      }
+    });
+  };
+
+  if (registration.installing) {
+    return listenInstalledStateChange();
+  }
+
+  // We are currently controlled so a new SW may be found...
+  // Add a listener in case a new SW is found,
+  registration.addEventListener('updatefound', listenInstalledStateChange);
+}
+
+window.addEventListener('load', () => {
+  // When the user asks to refresh the UI, we'll need to reload the window
+  navigator.serviceWorker.addEventListener('message', (event) => {
+    if (!event.data) {
+      return;
+    }
+
+    switch (event.data) {
+      case 'reload-window':
+        window.location.reload();
+        break;
+      default:
+        // NOOP
+        break;
+    }
+  });
+
+  navigator.serviceWorker.register('/sw.js')
+  .then(function (registration) {
+      // Track updates to the Service Worker.
+    if (!navigator.serviceWorker.controller) {
+      // The window client isn't currently controlled so it's a new service
+      // worker that will activate immediately
+      return;
+    }
+
+    onNewServiceWorker(registration, () => {
+      showRefreshUI(registration);
+    });
+  });
+});
+```
+
+This code handles the verious possible lifecycles of the service worker
+and detects when a new service worker has become installed and is waiting to
+activate.
+
+When a waiting service worker is found we set up a message listener on
+`navigator.serviceWorker` so we know when to reload the window. When the
+user clicks on the UI to refresh the page, we post a message to the new
+service worker telling it to force an activation. After this a message is
+posted to all windows telling them to reload.
+
+Note: This is one possible approach to refreshing the page on a new service
+worker. For a more thorough answer as well as an explanation of alternative
+approaches this
+[article by Redfin Engineering](https://redfin.engineering/how-to-fix-the-refresh-button-when-using-service-workers-a8e27af6df68)
+discuess a range of options.
+
+**Add to your service worker**
+
+```javascript
+self.addEventListener('message', (event) => {
+  if (!event.data){
+    return;
+  }
+
+  switch (event.data) {
+    case 'force-activate':
+      self.skipWaiting();
+      self.clients.claim();
+      self.clients.matchAll().then((clients) => {
+        clients.forEach((client) => client.postMessage('reload-window'));
+      });
+      break;
+    default:
+      // NOOP
+      break;
+  }
+});
+```
+
+This will receive a the 'force-activate' message and call `skipWaiting()` and
+`clients.claim()` so that the service worker activates immediately and controls
+all of the currently open windows, which in turn results in the `
+controllerchange` event firing causing the windows to reload.
```

Comment:
It's the `message` event, not the `controllerchange` event, that causes the `window`s to reload themselves.